utils/vlog/vlog_to_pbtype.py

```diff
@@ -55,294 +58,531 @@
 sys.path.insert(0, "..")
 from lib import xmlinc
 
-parser = argparse.ArgumentParser(
-    description=__doc__.strip(), formatter_class=argparse.RawTextHelpFormatter
+# $genblock$/vlog/tests/multiple_instance/multiple_instance.sim.v:12$64[57].\comb
+GENBLOCK_REGEX = re.compile(
+    "^\\$genblock\\$.*:[0-9]+\\$[0-9]+\\[(.*)\\]\\.\\\\(.*)"
 )
-parser.add_argument(
-    'infiles',
-    metavar='input.v',
-    type=str,
-    nargs='+',
-    help="""\
-One or more Verilog input files, that will be passed to Yosys internally.
-They should be enough to generate a flattened representation of the model,
-so that paths through the model can be determined.
-"""
-)
-parser.add_argument(
-    '--top',
-    help="""\
-Top level module, will usually be automatically determined from the file name
-%.sim.v
-"""
-)
-parser.add_argument(
-    '--includes',
-    help="""\
-Command seperate list of include directories.
-""",
-    default=""
-)
-parser.add_argument('-o', help="""\
-Output filename, default 'model.xml'
-""")
 
-args = parser.parse_args()
-iname = os.path.basename(args.infiles[0])
 
-outfile = "pb_type.xml"
-if "o" in args and args.o is not None:
-    outfile = args.o
+def strip_name(name: str, include_index=True) -> str:
+    """Convert generate block into normal array form.
+
+    >>> n = r"$genblock$/vlog/tests/multiple_instance/multiple_instance.sim.v:12$64[57].\comb"
+    >>> strip_name(n)
+    'comb[57]'
+    >>> strip_name(n, False)
+    'comb'
+    >>> n = r"$genblock$/vlog/tests/multiple_instance/multiple_instance.sim.v:15$10[3].\\comba"
+    >>> strip_name(n)
+    'comba[3]'
+    """
+    if name.startswith('$genblock$'):
+        index, name = GENBLOCK_REGEX.match(name).groups()
+        if include_index:
+            return "{}[{}]".format(name, index)
+        else:
+            return name
+    return name
 
-yosys.run.add_define("PB_TYPE")
-if args.includes:
-    for include in args.includes.split(','):
-        yosys.run.add_include(include)
 
-vjson = yosys.run.vlog_to_json(args.infiles, flatten=False, aig=False)
-yj = YosysJSON(vjson)
+CellName = str
+CellType = str
+PinName = str
+CellPin = Tuple[CellName, PinName]
 
-if args.top is not None:
-    top = args.top
-else:
-    wm = re.match(r"([A-Za-z0-9_]+)\.sim\.v", iname)
-    if wm:
-        top = wm.group(1).upper()
-    else:
-        print(
-            "ERROR file name not of format %.sim.v ({}), cannot detect top level. Manually specify the top level module using --top"
-            .format(iname)
-        )
-        sys.exit(1)
 
-tmod = yj.module(top)
+def create_port(
+        dir_xml: ET.Element, cell_pin: CellPin, direction: str, metadata=None
+) -> ET.Element:
```

Comment:
Docstrings are required throughout this file.

Reply:
@kgugala